opa/opa.go

```diff
@@ -0,0 +1,201 @@
+		DeniedStatusCode int
+
+		// DeniedResponseMessage response message when the request is denied.
+		DeniedResponseMessage string
+
+		// InputCreationMethod Function to generate the input.
+		InputCreationMethod InputCreationFunc
+
+		// ErrorHandler is custom callback to handle enforcing.
+		ErrorHandler func(c echo.Context, internal error, proposedStatus int) error
+	}
+
+	// InputCreationFunc is the function signature for functions that generate the input for the Rego query.
+	InputCreationFunc func(c echo.Context) (map[string]interface{}, error)
+)
+
+var (
+	// DefaultConfig is the default OPA middleware config.
+	DefaultConfig = Config{
+		
+		RegoQuery: "data.example.authz.allow",
+		Skipper: middleware.DefaultSkipper,
+		DeniedStatusCode: http.StatusBadRequest,
+		DeniedResponseMessage: "Request denied",
+		InputCreationMethod: defaultInput,
+		ErrorHandler: func(c echo.Context, internal error, proposedStatus int) error {
+			err := echo.NewHTTPError(proposedStatus, internal.Error())
+			err.Internal = internal
+			return err
+		},
+	}
+)
+
+// MiddlewareWithConfig returns a OPA middleware with config.
+// See `Middleware()`.
+func MiddlewareWithConfig(config Config) echo.MiddlewareFunc {
+	// fillAndValidate checks if any mandatory field is missing.
+	if err := config.fillAndValidate(); err != nil {
+		panic(err)
+	}
+
+	policy, err := ioutil.ReadAll(config.RegoPolicy)
+	if err != nil {
+		panic(fmt.Errorf("could not read rego policy: %w", err))
+	}
+
+	query, err := rego.New(
+		rego.Query(config.RegoQuery),
+		rego.Module("policy.rego", string(policy)),
+	).PrepareForEval(context.Background())
+
+	if err != nil {
+		panic(fmt.Errorf("rego policy error: %w", err))
+	}
+
+	return func(next echo.HandlerFunc) echo.HandlerFunc {
+		return func(c echo.Context) error {
+			if config.Skipper(c) {
+				return next(c)
+			}
+
+			input, err := config.InputCreationMethod(c)
+			if err != nil {
+				return config.ErrorHandler(c, err, http.StatusInternalServerError)
+			}
+
+			if config.IncludeQueryString {
+				input["query"] = c.QueryParams()
+			}
+
+			if len(config.IncludeHeaders) > 0 {
+				headers := make(map[string]string)
+				for _, header := range config.IncludeHeaders {
+					headers[header] = c.Request().Header.Get(header)
+				}
+				input["headers"] = headers
+			}
+
+			res, err := query.Eval(context.Background(), rego.EvalInput(input))
+			if err != nil {
+				return config.ErrorHandler(c, err, http.StatusInternalServerError)
+			}
+
+			if !res.Allowed() {
+				return config.ErrorHandler(c, errors.New("request denied"), config.DeniedStatusCode)
+			}
+
+			return next(c)
+		}
+	}
+}
+
+// fillAndValidate fills in default values for missing configuration fields and validates the configuration.
+func (c *Config) fillAndValidate() error {
+	if c.RegoQuery == "" {
+		return errors.New("rego query can not be empty")
+	}
+
+	if c.DeniedStatusCode == 0 {
+		c.DeniedStatusCode = DefaultConfig.DeniedStatusCode
```

Comment:
another remark about public global vars - `DefaultConfig.DeniedStatusCode` could be set to `0`